### ISSUE TITLE
iodine >0.7.1 set empty set-cookie headers, linux x86_64

### ISSUE THREAD

**Issue by ojab** [boazsegev/iodine#43]
With simple rack app 
```ruby
require 'sinatra'
require 'sinatra/cookies'

class App < Sinatra::Application # :nodoc:
  enable :sessions
  set :session_secret, 'my_secrets'

  get '/' do
    session[:user] = 'session_user'
    cookies['user'] = 'cookies_user'
    'Hello World!'
  end
end
```

I'm getting empty `Set-Cookie` headers using iodine > 0.7.1:
```
set-cookie:
set-cookie:
```
with version 0.7.1 I get correct
```
set-cookie: user=cookies_user; path=/; HttpOnly
set-cookie: rack.session=…; path=/; HttpOnly
```
`git bisect` told me that
```
ba5246923393d38f6e6a1be0374cd8f339f24459 is the first bad commit
commit ba5246923393d38f6e6a1be0374cd8f339f24459
Author: Bo <bo@bowild.com>
Date:   Wed Oct 3 04:38:23 2018 -0400

    facil.io updates

:040000 040000 0f4fab699832dcd8c23da77a1a8ee81f7db2338e b37a2f166fba3d836143f7aa1dd1fb48724a25e8 M      ext

git bisect start
# good: [73734170bb243b55d8ee8702ff9e67ebef641f03] fix for older gcc (heroko)
git bisect good 73734170bb243b55d8ee8702ff9e67ebef641f03
# bad: [1ae2f261cb24d36243821086f3827cbb00e378e9] docs
git bisect bad 1ae2f261cb24d36243821086f3827cbb00e378e9
# good: [21dc101c54b7dc188a0b2860c2d388da3102d3f0] on_idle scheduling
git bisect good 21dc101c54b7dc188a0b2860c2d388da3102d3f0
# bad: [f3029950252e239f3f295a307c730bcd814fc725] caught facil.io at a bad time...
git bisect bad f3029950252e239f3f295a307c730bcd814fc725
# skip: [4172f6cb807b2aa12f22361907bcb160c34dc697] facil.io updates
git bisect skip 4172f6cb807b2aa12f22361907bcb160c34dc697
# bad: [ba5246923393d38f6e6a1be0374cd8f339f24459] facil.io updates
git bisect bad ba5246923393d38f6e6a1be0374cd8f339f24459
# first bad commit: [ba5246923393d38f6e6a1be0374cd8f339f24459] facil.io updates
```

Linux x86_64 here (Fedora 29), please tell me if any other info is needed.

**Comment by boazsegev** [comment]
Hi Ojab,

Wow, I can't believe I didn't experience this issue before - thanks!

🙏🏻🙏🏻🙏🏻🙏🏻

I tracked it down and found out that when converting a String to an Array, the headers get zeroed out for an obscure and annoying reason.

I fixed the memory retention (reference counting) model in the HashMap and it seems to have fixed the issue (the `remove` function was, mistakingly, decreasing the reference count, something only `delete` should have been doing).

Anyway, I'm waiting for the CI tests to finish and I'll release an updated version as soon as I can.

Thanks again 🙏🏻🙏🏻🙏🏻
Bo.

**Comment by boazsegev** [comment]
I just released iodine 0.7.7 with a fix for this issue.

Thank you again for exposing this critical concern 🙏🏻🙏🏻🙏🏻👏🏻👍🏻

Kindly,
   Bo.

P.S.

You probably know this, but I thought I might point out that Sinatra isn't as "light" as you might think. It's not really a simple rack app.

Using just Rack to recreated the issue, iodine managed 41,334 Req/Sec (with logging enabled), while using Sinatra the performance dropped to 2,943 Req/Sec (I noticed this while testing for memory leaks)... not that it would matter once database access and routing are required, but FYI.
